File: examples/showcase/export-screening/evals/validate_risk_output.py
#!/usr/bin/env python3
"""
Export Risk Output Validator for AgentV

Validates that the candidate answer is valid JSON with required fields,
and extracts the risk classification for confusion matrix computation.

Returns structured output that enables post-processing for metrics.
"""

import json
import sys
from typing import Any


VALID_RISK_LEVELS = {"High", "Medium", "Low"}
REQUIRED_KEYS = ["riskLevel", "reasoning"]


def extract_json_from_response(content: str) -> dict[str, Any] | None:
    """Extract JSON from response, handling markdown code fences."""
    content = content.strip()

    # Handle markdown code fences
    if content.startswith("```"):
        lines = content.split("\n")
        lines = lines[1:]  # Remove opening fence
        if lines and lines[-1].strip() == "```":
            lines = lines[:-1]  # Remove closing fence
        content = "\n".join(lines).strip()

    try:
        return json.loads(content)
    except json.JSONDecodeError:
        return None


def extract_expected_risk_level(expected_messages: list[dict]) -> str | None:
    """Extract expected risk level from expected_messages."""
    if not expected_messages:
        return None

    for msg in expected_messages:
        if msg.get("role") != "assistant":
            continue
        content = msg.get("content")
        if isinstance(content, dict) and "riskLevel" in content:
            return content["riskLevel"]
        if isinstance(content, str):
            try:
                parsed = json.loads(content)
                if isinstance(parsed, dict) and "riskLevel" in parsed:
                    return parsed["riskLevel"]
            except json.JSONDecodeError:
                pass
    return None


def validate_risk_output(
    candidate_answer: str,
    expected_messages: list[dict] | None = None
) -> dict[str, Any]:
    """
    Validate risk assessment output and compare to expected.

    Returns structured result for AgentV evaluation.
    """
    hits: list[str] = []
    misses: list[str] = []

    # Parse candidate JSON
    parsed = extract_json_from_response(candidate_answer)

    if parsed is None:
        return {
            "score": 0.0,
            "hits": [],
            "misses": ["Output is not valid JSON"],
            "reasoning": "Failed to parse response as JSON"
        }

    if not isinstance(parsed, dict):
        return {
            "score": 0.0,
            "hits": [],
            "misses": [f"Output is {type(parsed).__name__}, expected object"],
            "reasoning": "Response must be a JSON object"
        }

    # Check required keys
    missing_keys = [k for k in REQUIRED_KEYS if k not in parsed]
    if missing_keys:
        return {
            "score": 0.0,
            "hits": [],
            "misses": [f"Missing required keys: {', '.join(missing_keys)}"],
            "reasoning": f"Response missing: {', '.join(missing_keys)}"
        }

    hits.append("Valid JSON with required keys")

    # Validate riskLevel value
    candidate_risk = parsed.get("riskLevel")
    if candidate_risk not in VALID_RISK_LEVELS:
        misses.append(f"Invalid riskLevel: '{candidate_risk}' (must be High/Medium/Low)")
        return {
            "score": 0.25,
            "hits": hits,
            "misses": misses,
            "reasoning": f"riskLevel '{candidate_risk}' is not valid"
        }

    hits.append(f"riskLevel={candidate_risk}")

    # Compare to expected if available
    expected_risk = None
    if expected_messages:
        expected_risk = extract_expected_risk_level(expected_messages)

    if expected_risk is None:
        # No expected value to compare - just validate format
        return {
            "score": 1.0,
            "hits": hits,
            "misses": misses,
            "reasoning": f"Valid response with riskLevel={candidate_risk}"
        }

    # Classification comparison
    if candidate_risk == expected_risk:
        hits.append(f"Correct: AI={candidate_risk}, Expected={expected_risk}")
        score = 1.0
        reasoning = f"Correctly classified as {candidate_risk}"
    else:
        misses.append(f"Mismatch: AI={candidate_risk}, Expected={expected_risk}")
        score = 0.0
        reasoning = f"Misclassified: AI={candidate_risk}, Expected={expected_risk}"

    return {
        "score": score,
        "hits": hits,
        "misses": misses,
        "reasoning": reasoning
    }


def main():
    """Main entry point for AgentV code evaluator."""
    try:
        eval_data = json.load(sys.stdin)
    except json.JSONDecodeError as e:
        print(json.dumps({
            "score": 0.0,
            "hits": [],
            "misses": [f"Failed to parse evaluator input: {e}"],
            "reasoning": "Internal error parsing eval input"
        }))
        sys.exit(1)

    candidate_answer = eval_data.get("candidate_answer", "")
    expected_messages = eval_data.get("expected_messages")

    result = validate_risk_output(candidate_answer, expected_messages)

    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: examples/showcase/export-screening/evals/validate_risk_output.ts
#!/usr/bin/env bun
/**
 * Export Risk Output Validator for AgentV
 *
 * Validates that the candidate output is valid JSON with required fields,
 * and extracts the risk classification for confusion matrix computation.
 *
 * Returns structured output that enables post-processing for metrics.
 */
import { defineCodeGrader } from '@agentv/eval';

const VALID_RISK_LEVELS = new Set(['High', 'Medium', 'Low']);
const REQUIRED_KEYS = ['riskLevel', 'reasoning'];

function extractJsonFromResponse(content: string): Record<string, unknown> | null {
  let trimmed = content.trim();

  // Handle markdown code fences
  if (trimmed.startsWith('```')) {
    const lines = trimmed.split('\n');
    lines.shift(); // Remove opening fence
    if (lines.length > 0 && lines[lines.length - 1].trim() === '```') {
      lines.pop(); // Remove closing fence
    }
    trimmed = lines.join('\n').trim();
  }

  try {
    const parsed = JSON.parse(trimmed);
    return typeof parsed === 'object' && parsed !== null ? parsed : null;
  } catch {
    return null;
  }
}

function extractExpectedRiskLevel(
  expectedOutput: readonly Record<string, unknown>[] | undefined,
): string | null {
  if (!expectedOutput) return null;

  for (const msg of expectedOutput) {
    if (msg.role !== 'assistant') continue;

    const content = msg.content;
    if (typeof content === 'object' && content !== null && 'riskLevel' in content) {
      return (content as { riskLevel: string }).riskLevel;
    }
    if (typeof content === 'string') {
      try {
        const parsed = JSON.parse(content);
        if (typeof parsed === 'object' && parsed !== null && 'riskLevel' in parsed) {
          return parsed.riskLevel;
        }
      } catch {
        // Ignore parse errors
      }
    }
  }
  return null;
}

function getMessageText(
  messages: readonly { role: string; content?: unknown }[],
  role = 'assistant',
): string {
  for (let i = messages.length - 1; i >= 0; i--) {
    const msg = messages[i];
    if (msg.role === role) {
      if (typeof msg.content === 'string') return msg.content;
      if (Array.isArray(msg.content)) {
        return msg.content
          .filter((b: { type?: string }) => b.type === 'text')
          .map((b: { text?: string }) => b.text)
          .join('\n');
      }
    }
  }
  return '';
}

export default defineCodeGrader(({ output, expectedOutput }) => {
  const outputText = getMessageText(output ?? []);
  const assertions: Array<{ text: string; passed: boolean; evidence?: string }> = [];

  // Parse candidate JSON
  const parsed = extractJsonFromResponse(outputText);

  if (parsed === null) {
    return {
      score: 0.0,
      assertions: [
        {
          text: 'Output is not valid JSON',
          passed: false,
          evidence: 'Failed to parse response as JSON',
        },
      ],
    };
  }

  // Check required keys
  const missingKeys = REQUIRED_KEYS.filter((k) => !(k in parsed));
  if (missingKeys.length > 0) {
    return {
      score: 0.0,
      assertions: [{ text: `Missing required keys: ${missingKeys.join(', ')}`, passed: false }],
    };
  }

  assertions.push({ text: 'Valid JSON with required keys', passed: true });

  // Validate riskLevel value
  const candidateRisk = parsed.riskLevel as string;
  if (!VALID_RISK_LEVELS.has(candidateRisk)) {
    assertions.push({
      text: `Invalid riskLevel: '${candidateRisk}' (must be High/Medium/Low)`,
      passed: false,
    });
    return {
      score: 0.25,
      assertions,
    };
  }

  assertions.push({ text: `riskLevel=${candidateRisk}`, passed: true });

  // Compare to expected if available
  const expectedRisk = extractExpectedRiskLevel(expectedOutput);

  if (expectedRisk === null) {
    // No expected value to compare - just validate format
    return {
      score: 1.0,
      assertions,
    };
  }

  // Classification comparison
  if (candidateRisk === expectedRisk) {
    assertions.push({
      text: `Correct: AI=${candidateRisk}, Expected=${expectedRisk}`,
      passed: true,
    });
    return {
      score: 1.0,
      assertions,
    };
  }

  assertions.push({
    text: `Mismatch: AI=${candidateRisk}, Expected=${expectedRisk}`,
    passed: false,
  });
  return {
    score: 0.0,
    assertions,
  };
});
